test/docker-e2e/e2e_test.go

```diff
@@ -171,3 +187,256 @@ func (s *CelestiaTestSuite) WaitForSync(ctx context.Context, statusClient rpccli
+	}
+
+	endValidators, err := rpcClient.Validators(ctx, &endHeight, nil, nil)
+	if err != nil {
+		return nil, nil, fmt.Errorf("validators query at end height %d: %w", endHeight, err)
+	}
+
+	return startValidators, endValidators, nil
+}
+
+// fetchProposerAddresses gathers proposer addresses from block headers using efficient batching
+func (s *CelestiaTestSuite) fetchProposerAddresses(
+	ctx context.Context,
+	rpcClient rpcclient.Client,
+	startHeight, endHeight int64,
+) ([]string, error) {
+	var (
+		proposers = make(map[string]struct{})
+
+		// Check blocks after startHeight (exclusive) through endHeight (inclusive)
+		minHeight = startHeight + 1
+		maxHeight = endHeight
+	)
+
+	for minHeight <= maxHeight {
+		// BlockchainInfo returns headers in descending order and may limit the range
+		blockchainInfo, err := rpcClient.BlockchainInfo(ctx, minHeight, maxHeight)
+		if err != nil {
+			return nil, fmt.Errorf("failed to fetch block headers for range %d-%d: %w", minHeight, maxHeight, err)
+		}
+
+		if len(blockchainInfo.BlockMetas) == 0 {
+			return nil, fmt.Errorf("no block headers returned for range %d-%d", minHeight, maxHeight)
+		}
+
+		for _, blockMeta := range blockchainInfo.BlockMetas {
+			addr := blockMeta.Header.ProposerAddress.String()
+			proposers[addr] = struct{}{}
+		}
+
+		// Continue from the lowest height we just processed
+		// BlockMetas are in descending order, so the last one has the lowest height
+		lastProcessedHeight := blockchainInfo.BlockMetas[len(blockchainInfo.BlockMetas)-1].Header.Height
+
+		// If we've processed all requested heights, break
+		if lastProcessedHeight <= minHeight {
+			break
+		}
+
+		// Move to the next batch
+		maxHeight = lastProcessedHeight - 1
+	}
+
+	// Convert map to slice for cleaner return type
+	result := make([]string, 0, len(proposers))
+	for addr := range proposers {
+		result = append(result, addr)
+	}
+
+	blocksProduced := endHeight - startHeight
+	s.T().Logf("Found %d unique proposers across %d blocks", len(result), blocksProduced)
+	return result, nil
+}
+
+// validateAllValidatorsProposed ensures every validator proposed at least one block
+func (s *CelestiaTestSuite) validateAllValidatorsProposed(
+	startValidators, endValidators *coretypes.ResultValidators,
+	proposerAddresses []string,
+	startHeight, endHeight int64,
+) error {
+	// Convert slice to map for O(1) lookups
+	proposers := make(map[string]struct{}, len(proposerAddresses))
+	for _, addr := range proposerAddresses {
+		proposers[addr] = struct{}{}
+	}
+
+	// Create a combined map of all validators that should have proposed
+	var (
+		allValidators  = make(map[string]struct{})
+		validatorNames = make(map[string]string) // for better error reporting
+	)
+
+	// Add start validators
+	for _, val := range startValidators.Validators {
+		addr := val.Address.String()
+		allValidators[addr] = struct{}{}
+		validatorNames[addr] = fmt.Sprintf(validatorNameFormat, addr[:validatorAddrTruncateLen]) // shortened for readability
```

Comment:
[nit] can we eliminate the separate `validatorNames` map by just using the entire address in `allValidators`? If yes that seem desirable to reduce complexity (maintaining multiple maps) and unnecessary constants (`validatorAddrTruncateLen`)

Reply:
technically we can do that, it is only used to format error messages, but we can eliminate it entirely and just use the raw addresses directly.